fabtests/functional/bw.c

```diff
@@ -0,0 +1,244 @@
+			    opts.transfer_size);
+
+	return ft_post_tx_buf(ep, remote_fi_addr, opts.transfer_size,
+			      NO_CQ_DATA, &msg->context, msg->buf,
+			      msg->desc, 0);
+}
+
+static ssize_t wait_check_rx_bufs(void)
+{
+	ssize_t ret;
+	int i;
+
+	ret = ft_get_rx_comp(rx_seq);
+	if (ret)
+		return ret;
+
+	if (!ft_check_opts(FT_OPT_VERIFY_DATA | FT_OPT_ACTIVE))
+		return 0;
+
+	for (i = 0; i < opts.window_size; i++) {
+		ret = ft_check_buf((char *) rx_ctx_arr[i].buf +
+				   ft_rx_prefix_size(), opts.transfer_size);
+		if (ret)
+			return ret;
+	}
+
+	return 0;
+}
+
+static int post_rx_sync(void)
+{
+	int ret;
+
+	ret = ft_post_rx(ep, rx_size, &rx_ctx);
+	if (ret)
+		return ret;
+
+	if (opts.dst_addr) {
+		ret = ft_tx(ep, remote_fi_addr, 1, &tx_ctx);
+		if (ret)
+			return ret;
+
+		ret = ft_get_rx_comp(rx_seq);
+	} else {
+		ret = ft_get_rx_comp(rx_seq);
+		if (ret)
+			return ret;
+
+		ret = ft_tx(ep, remote_fi_addr, 1, &tx_ctx);
+	}
+
+	return ret;
+}
+
+static int run_loop(void)
+{
+	int ret, i;
+
+	/* Receive side delay is used in order to let the sender
+ 	   get ahead of the receiver and post multiple sends
+	   before the receiver begins processing them. */
+	if (!opts.dst_addr)
+		sleep(sleep_time);
+
+	ft_start();
+	if (opts.dst_addr) {
+		for (i = 0; i < opts.window_size; i++) {
+			ret = post_one_tx(&tx_ctx_arr[i]);
+			if (ret)
+				return ret;
+		}
+
+		ret = ft_get_tx_comp(tx_seq);
+		if (ret)
+			return ret;
+	} else {
+		for (i = 0; i < opts.window_size; i++) {
+			ret = ft_post_rx_buf(ep, opts.transfer_size,
+					     &rx_ctx_arr[i].context,
+					     rx_ctx_arr[i].buf,
+					     rx_ctx_arr[i].desc, 0);
+			if (ret)
+				return ret;
+		}
+
+		ret = wait_check_rx_bufs();
+		if (ret)
+			return ret;
+	}
+	ft_stop();
+
+	if (opts.options & FT_OPT_OOB_SYNC)
+		ret = ft_sync();
+	else
+		ret = post_rx_sync();
+	if (ret)
+		return ret;
+
+	if (opts.machr)
+		show_perf_mr(opts.transfer_size, opts.window_size, &start, &end, 1,
+				opts.argc, opts.argv);
+	else
+		show_perf(NULL, opts.transfer_size, opts.window_size, &start, &end, 1);
+
+	return ret;
+}
+
+static int run(void)
+{
+	int ret, i;
+
+	ret = hints->ep_attr->type == FI_EP_MSG ?
+		ft_init_fabric_cm() : ft_init_fabric();
+	if (ret)
+		return ret;
+	
+	ret = ft_tx(ep, remote_fi_addr, 1, &tx_ctx);
+	if (ret)
+		return ret;
+
+	ret = ft_get_tx_comp(tx_seq);
+	if (ret)
+		return ret;
+
+	ret = ft_get_rx_comp(rx_seq);
+	if (ret)
+		return ret;
+
+	if (!(opts.options & FT_OPT_SIZE)) {
+		for (i = 0; i < TEST_CNT; i++) {
+			if (!ft_use_size(i, opts.sizes_enabled))
+				continue;
+			opts.transfer_size = test_size[i].size;
+			ret = run_loop();
+			if (ret)
+				goto out;
+		}
+	} else {
+		ret = run_loop();
+		if (ret)
+			goto out;
+	}
+
+out:
+	return ret;
+}
+
+int main(int argc, char **argv)
+{
+	int op, ret;
+
+	opts = INIT_OPTS;
+
+	hints = fi_allocinfo();
+	if (!hints)
+		return EXIT_FAILURE;
+
+	hints->ep_attr->type = FI_EP_RDM;
+
```

Comment:
Without setting a msg_order hint, EFA will not reorder packets, which might explain the validation error.

Reply:
Pulled down the patch and added msg_order, and that let the test pass with EFA:

```
fi_bw -e rdm -v -T 1 -p "efa":                                              Pass
fi_bw -e msg -v -T 1 -p "efa":                                          Excluded
```

Reply:
This is a good thing.  With the per request context, we can support out of order transfers, but still provide data validation.  We just need to figure out the starting data based on the context corresponding to the completion.  I would avoid adding a msg_order requirement here and update the data set/check calls to adjust based on the index of the message buffer.